Reblog feature: small customization of site picker

develric opened the pull request:
This PR tries to further enhance the #11555 by adding a little step in site picker selection before to continue the reblog flow when a user has more than one site.

| NOTE | IMAGE |
|---|---|
| Adding `CONTINUE` menu disabled at beginning | ![image](https://user-images.githubusercontent.com/47797566/79810282-fcdbe000-8372-11ea-950b-f28e9cc6cdc4.png) |
| Enabling `CONTINUE` menu after first site selection  | ![image](https://user-images.githubusercontent.com/47797566/79810457-6360fe00-8373-11ea-9af9-a02217be746e.png) |
| Search will allow site filtering and on filtered site selection user will be again in a  screen like the first one above  | ![image](https://user-images.githubusercontent.com/47797566/79810850-5abcf780-8374-11ea-8783-f59d591f747c.png)  |

cc @osullivanchris and @iamthomasbishop 

## To test
Be sure to be logged in with a WP.com account having more than one site (you should have enough sites to need to scroll in the site picker so to manual test the scroll to selected scenario)

### Feature Test
- From a reader post select the reblog button
- Check you get access to the site picker like below without any selected site

<p align="center">
  <img width="300" src=https://user-images.githubusercontent.com/47797566/80132154-8ec63180-859b-11ea-9c7f-338fb36417d3.png>
</p>

- Check single site is selected if you tap on it, and you get an action bar like below

<p align="center">
  <img width="300" src=https://user-images.githubusercontent.com/47797566/80132107-7fdf7f00-859b-11ea-9de5-aab7ce257b8e.png>
</p>

- Check you can select other sites and still remain in the same screen as above with an updated action bar title set to the site title
- Check you can go back and come to the initial screen with no site selected
- Long press on a site and check nothing happens (hiding sites capability should not be available)
- Activate search and type some search string to get a filtered list of sites
- Select a site that is at the lower end of the main list
- Check the search view is dismissed and the sites list is scrolled to the selected site
- Tap on the Checkmark icon and check you can complete the reblog flow.

### Smoke Test previous available scenarios
Site Picker is used in `My Site > Switch Site`, `Share to WP`, `Login Epilogue`; Smock test these scenarios to check for eventual regressions. In particular check the Site Hide feature in `Switch Site` picker works as usual.


PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

iamthomasbishop commented:
@develric I haven't had a chance to download a test build, but looking at the screenshots a couple of things stand out to me, some of which @osullivanchris has already mentioned.

* The `Continue` button def feels a little off. I think a `checkmark` icon might be a bit more balanced. The trade-off is that it might be _slightly_ less obvious, although translations on that `Continue` label could also make it feel less balanced if longer. Another option for the label (to keep it short) could be `Go`.
* I think the selection state on the list row could be a bit stronger. I'm not sure if that's the standard/generic style that we are inheriting or a custom, but I wonder if we should make the highlight stronger or add an icon/indicator when a list item is selected?

develric commented:
Hi @osullivanchris and @iamthomasbishop , thanks for the follow up 😊!

> Perhaps we could scroll to the position of the selected site?

Agree and modified the code (you should be able to check using the new APK from the above link)

> The Continue button def feels a little off...

Below how it results in both version; I still feel `continue` sounds clearer at least in english but fully get your concerns and probably would go with the checkmark icon; let me know wdyt, thanks 😊.

| ![image](https://user-images.githubusercontent.com/47797566/79919416-a502af00-842e-11ea-90c7-6953d0e3b7e0.png) | ![image](https://user-images.githubusercontent.com/47797566/79919448-b8157f00-842e-11ea-8790-23769e1bc2c8.png) |
|---|---|
| ![image](https://user-images.githubusercontent.com/47797566/79919962-9963b800-842f-11ea-962a-c50bb0c9c3fd.png) | ![image](https://user-images.githubusercontent.com/47797566/79920034-ca43ed00-842f-11ea-9d83-5277ffbb6e15.png) |


> I'm not sure if that's the standard/generic style that we are inheriting or a custom...

Yep, I used (only customized some behavior) the same site picker we have when switching site in My Site (see below image), also because this one has the search already in place like in iOS. 

<p align="center">
  <img width="300" src=https://user-images.githubusercontent.com/47797566/79920392-50603380-8430-11ea-9b3d-6c8548d595c5.png>
</p>

My understanding was we were trying to minimize modifications here but let me know 😊 (some site picker revisiting could be a good thing to tackle anyway, maybe in a dedicated task? 🤷‍♂️)

osullivanchris commented:
Thanks for the fixes @develric ! The checkmark looks better, and that's great you were able to do the scrolling thing easily enough.

> My understanding was we were trying to minimize modifications here but let me know 😊 (some site picker revisiting could be a good thing to tackle anyway, maybe in a dedicated task? 🤷‍♂️)

I agree this is just a generic/global thing we're inheriting here, so I don't think it needs to be addressed as part of this task

develric commented:
Thanks @osullivanchris , confirming final status as per below images:

| ![image](https://user-images.githubusercontent.com/47797566/79989098-be980b00-84af-11ea-88de-6fc860795dd6.png) | ![image](https://user-images.githubusercontent.com/47797566/79989125-ca83cd00-84af-11ea-89ed-cf448497f403.png) |
|---|---|

iamthomasbishop commented:
@develric @osullivanchris Checkmark definitely feels better, although now that I see it I have a suggestion — and I apologize for not catching this earlier in the process, I should've thought of it earlier. Would it make more sense to perhaps morph the app bar into a "selection mode" app bar — in other words, something like this:

<a href="https://user-images.githubusercontent.com/1200432/80003535-cb543900-8486-11ea-9dc0-272e24108307.jpeg"><img src="https://user-images.githubusercontent.com/1200432/80003535-cb543900-8486-11ea-9dc0-272e24108307.jpeg" width="700"></a>

| **No selection** | **Site selected** |
| --------------- | --------------- |
| `X` `Choose site` `🔍` | `<-` `[Site name]` `√` |

I think this would make selection even clearer and better align with Material. I do think the solution proposed in the last comment is workable, but I think this would be clearer, cleaner, and achieve the same result.

osullivanchris commented:
That is very nice @iamthomasbishop 

are we using that pattern anywhere else?

iamthomasbishop commented:
> are we using that pattern anywhere else?

@osullivanchris Good question. I hadn't even realized this, but we're using the same pattern already (diff actions, but same concept) on the site picker that shows when you tap on the `Switch site` button on the My Site view (when you get to the picker, long-press on a site cell).

develric commented:
Hi @iamthomasbishop and @osullivanchris 👋! I updated the code and now it should look (and hopefully behave 😊) as described by Thomas. You should be able to download the apk, below some screenshots for quick reference.

| ![image](https://user-images.githubusercontent.com/47797566/80133084-fb8dfb80-859c-11ea-8232-67f8355836c8.png) | ![image](https://user-images.githubusercontent.com/47797566/80133119-09438100-859d-11ea-8ff5-3eb7fd006e37.png) |
|---|---|




**Note for the code reviewer:** I updated the test steps in the PR description but not the initial images for history purposes. In the code I tried to move logic into the View Model I added but still some logic is in between the Activity/Adapter and the VM. I'm not trilled with that but I was trying to minimize the changes (as said previously probaby a review of the site pickers could make sense, for example screen rotation is without some memory of the status; I tried to cover it at least in this reblog scenario). Let me know 😊.

osullivanchris commented:
Thanks for the steps in the description the top, really helpful to make sure I'm covering everything.

This works really well, I really like it, great job. Glad we were able to make some nice improvements here.

The only tiny thing I can think of is
> Check you can go back and come to the initial screen with no site selected

I wonder should tapping the selected site again deselect it?

Its such a minor thing so I'm fine without it also! Thanks @iamthomasbishop  for the great suggestion on the action bar

develric commented:
Following up after a quick chat on slack with @osullivanchris (thanks for your time 🙇‍♂️)

> I wonder should tapping the selected site again deselect it?

Our current understanding is that it's not worth changing (I also think we could aim to some review of the app pickers) and we can proceed with the code review (pinging @khaykov and @planarvoid to let them know).

iamthomasbishop commented:
> I wonder should tapping the selected site again deselect it?

@osullivanchris I think I agree, although it should prob be done as a separate PR as @develric suggested.

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -226,7 +226,7 @@
         EditPostSettingsCallback,
         PrivateAtCookieProgressDialogOnDismissListener {
-    public static final String ACTION_REBLOG = "reblogAction";		
+    public static final String ACTION_REBLOG = "reblogAction";
```
I think there are no real changes to this file so you can revert it :-)

develric replied:
yep no real changes, but AS was complaining about trailing tabs so I removed them; if you agree I would keep this, let me know.

develric commented on WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerActivity.java:
```diff
@@ -116,6 +120,60 @@ public void onCreate(Bundle savedInstanceState) {
+                            );
+                            ToastUtils.showToast(this, R.string.site_picker_ask_site_select);
+                        }
+
```
ehm, not actually 😊..removed.

develric commented:
Thanks for the review and spotting that issue @planarvoid 🙇‍♂️. Should have fixed it (bad copy paste error 🤦‍♂️).

This PR started as a "small customisation" of the site picker (as the title still says 😊) but ended up being more complex. Coming back I would probably create a dedicated picker for the reblog task but as said maybe could make sense to review the site picker component instead outside of this task. Let me know wdyt 🙇‍♂️.
